### PR TITLE
[rllib] Parallel-data loading and multi-gpu support for IMPALA

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

- Add support for multi-gpu optimizer to IMPALA.
- Add option for parallel data loading and replay.
- These are all disabled by default.

### CODE REVIEW

**Comment by richardliaw** [comment]
Is this still WIP or ready for review? also conflicts.

**Comment by ericl** [comment]
This is now ready for review.

File: python/ray/rllib/optimizers/async_samples_optimizer.py [hunk 1]
```diff
@@ -48,32 +54,173 @@ def run(self):
...
 
+class TFMultiGPULearner(LearnerThread):
+    def __init__(self,
+                 local_evaluator,
+                 num_gpus=2,
+                 lr=0.0005,
+                 train_batch_size=500,
+                 grad_clip=40,
+                 num_parallel_data_loaders=1):
+        import tensorflow as tf
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_samples_optimizer.py]: can you note why this is here?

File: python/ray/rllib/optimizers/async_samples_optimizer.py [hunk 2]
```diff
@@ -48,32 +54,173 @@ def run(self):
...
-                self.weights_updated += 1
-                if "stats" in fetches:
-                    self.stats = fetches["stats"]
-            self.outqueue.put(batch.count)
+            batch = self.inqueue.get()
+
+        with self.grad_timer:
+            fetches = self.local_evaluator.compute_apply(batch)
+            self.weights_updated = True
+            if "stats" in fetches:
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_samples_optimizer.py]: `self.stats = fetches.get("stats", {})`?

File: python/ray/rllib/optimizers/async_samples_optimizer.py [hunk 3]
```diff
@@ -48,32 +54,173 @@ def run(self):
...
+
+    def step(self):
+        assert self.loader_thread.is_alive()
+        with self.load_wait_timer:
+            opt = self.ready_optimizers.get()
+
+        with self.grad_timer:
+            fetches = opt.optimize(self.sess, 0)
+            self.weights_updated = True
+            if "stats" in fetches:
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_samples_optimizer.py]: `self.stats = fetches.get("stats", {})`?

File: python/ray/rllib/optimizers/async_samples_optimizer.py [hunk 4]
```diff
@@ -48,32 +54,173 @@ def run(self):
...
+
+    def run(self):
+        while True:
+            self.step()
+
+    def step(self):
+        l = self.learner
+        with self.queue_timer:
+            batch = l.inqueue.get()
+            assert batch.count == l.train_batch_size, batch.count
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_samples_optimizer.py]: maybe print both batch.count and l.train_batch_size?

File: python/ray/rllib/optimizers/async_samples_optimizer.py [hunk 5]
```diff
@@ -48,32 +54,173 @@ def run(self):
...
+                self.stats = fetches["stats"]
+
+        self.outqueue.put(batch.count)
         self.learner_queue_size.push(self.inqueue.qsize())
 
 
+class TFMultiGPULearner(LearnerThread):
+    def __init__(self,
+                 local_evaluator,
+                 num_gpus=2,
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_samples_optimizer.py]: why default this to 2?

File: python/ray/rllib/optimizers/async_samples_optimizer.py [hunk 6]
```diff
@@ -48,32 +54,173 @@ def run(self):
...
+            fetches = self.local_evaluator.compute_apply(batch)
+            self.weights_updated = True
+            if "stats" in fetches:
+                self.stats = fetches["stats"]
+
+        self.outqueue.put(batch.count)
         self.learner_queue_size.push(self.inqueue.qsize())
 
 
+class TFMultiGPULearner(LearnerThread):
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_samples_optimizer.py]: Can you add some class documentation?

**Review by richardliaw — COMMENTED** [review]
Overall looks good; left couple small comments. A couple notes - 

1. Any better way of unit testing these components? Especially the optimizer.
2. I would argue that despite this being internal code, documenting these classes would benefit this project a lot. Also, having public documentation for the shared configs and algorithm specific things would be really good.

**Review by ericl — COMMENTED** [review]
Personally, I think the multi-GPU stuff is too messy and we should move away from it. However it's not clear what the best way to do that is, perhaps with TF eager, or moving a lot of the logic into numpy, it would be simpler.

**Comment by ericl** [comment]
I just did a quick multi-GPU run on PPO / IMPALA atari and perf looks good still.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**